streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala

```diff
@@ -223,22 +227,40 @@ class ReceiverTracker(ssc: StreamingContext, skipReceiverLaunch: Boolean = false
       thread.start()
     }
 
-    def stop() {
+    def stop(graceful: Boolean) {
       // Send the stop signal to all the receivers
       stopReceivers()
 
       // Wait for the Spark job that runs the receivers to be over
       // That is, for the receivers to quit gracefully.
+      val stopTimeout = conf.getLong(
+        "spark.streaming.gracefulStopTimeout",
+        10 * ssc.graph.batchDuration.milliseconds)
+
+
       thread.join(10000)
 
       // Check if all the receivers have been deregistered or not
-      if (!receiverInfo.isEmpty) {
-        logWarning("All of the receivers have not deregistered, " + receiverInfo)
+      def done = { receiverInfo.isEmpty && terminated }
+
+      if (graceful) {
+        TimeoutUtils.waitUntilDone(stopTimeout,() => done ) match {
```

Comment:
Kind of a toss-up over whether to use `match` vs `if` here, I guess.  Matching on a boolean feels a bit strange to me.